### ISSUE TITLE
Cluster API should not use controller refs for resources that it does not create

### ISSUE THREAD

**Issue by detiber** [kubernetes-sigs/cluster-api#4014]
**What steps did you take and what happened:**

Attempting to create a Cluster using an external tool such as Crossplane causes issues with both CAPI and Crossplane wanting to have the singular controller reference for resources that are created from Crossplane.

We've used controller refs in the past to avoid the possibility of multiple Clusters owning the same resources, but siince this hampers interoperability, the use of non-controller ownerReferences should be used and the verification that a resource isn't "owned" by the same cluster should be handled separately.

**What did you expect to happen:**

Crossplane should be able to create a Cluster API Cluster

**Anything else you would like to add:**

Anywhere that Cluster API controllers are creating resources should still continue to use controller references.

/kind bug

**Comment by hasheddan** [comment]
@vincepri I can give some more insight here, and would also love to attend a community meeting to present some ideas / demos. CAPI and Crossplane can interact at two main levels:
1. CAPI could rely on Crossplane providers to supply the backing infrastructure for cloud providers as Kubernetes objects. Crossplane providers are similar to the CAPI infrastructure providers except they are more general purpose and support all types across a provider rather than just those needed to spin up a k8s cluster.
2. Crossplane can orchestrate CAPI types using [composition](https://crossplane.io/docs/v0.14/getting-started/compose-infrastructure.html). You can see an _extremely rough_ sketch of what this could look like in a [demo we are putting together](https://github.com/crossplane-contrib/cross-capi-metal/blob/ac161ba28e050b353188e12620ca567d77f267b8/package/capi/composition.yaml). This is the area in which we are running into issues as Crossplane's composition engine adds some controller references to the objects it composes, which CAPI is relying on controlling for its operations. We specifically hit this with the `PacketCluster` needing to be controlled by the CAPI `Cluster` type.

The `Cluster` is kind of an interesting case because I would typically not put controller references on a user-created object (i.e. the `PacketCluster` is part of the CAPI quickstart output that a user applies, or in this case, Crossplane applies), but I understand the rationale for it here. The controller refs are just one example of how Crossplane and CAPI can encounter situations where they don't play nicely, but I was actually pleasantly surprised that there were fewer conflicts than I was expecting.

I think a broader discussion about how CAPI and Crossplane can integrate would be beneficial here, and I am more than happy to create some content that can be presented. In many ways, the components of CAPI are similar to many Crossplane providers:
- The Kubeadm bootstrap provider follows a similar model to projects like [provider-helm](https://github.com/crossplane-contrib/provider-helm) and [provider-sql](https://github.com/crossplane-contrib/provider-sql)
- The various infrastructure providers are similar to their Crossplane counterparts (e.g. [provider-gcp](https://github.com/crossplane/provider-gcp), [provider-aws](https://github.com/crossplane/provider-aws), etc.)
- The core CAPI functionality is akin to a more opiononated Crossplane [composition engine](https://crossplane.io/docs/v0.14/introduction/composition.html)

**Comment by fabriziopandini** [comment]
/milestone v0.4.0
given that some investigation work is going on

**Comment by fabriziopandini** [comment]
@hasheddan we are trying to close API changes for the v1alpha4 version.
Are there some some actionable outcome from your investigation?

/priority important-soon

**Comment by hasheddan** [comment]
@fabriziopandini yes, there has been significant work around using crossplane as a drop in replacement for types in cluster-api-provider-gcp. The work is being tracked in https://github.com/crossplane/provider-gcp/issues/303 and most of it has taken place on [livestreams that have been recorded](https://youtube.com/playlist?list=PL510POnNVaaazaBA7jNh1f6Qz1LxuB5Ku). I would love to give a demo to CAPI community soon (potentially next week) -- what would you recommend as best forum for that?

Note that I _believe_ that any crossplane integration will not need to block `v1alpha4` API changes as one of the benefits of the model is that we can mold the composite resources (which are the drop-in types for existing CAPI provider types) to fit any API.

/cc @vincepri

**Comment by vincepri** [comment]
I think for now we can delay a change in references until we find a suitable approach that works across all the proposed integrations, for now we can possibly use what's in #4135 to have a crossplane integration

/milestone Next

**Comment by lukebond** [comment]
/reopen

i'd like to reopen this issue as originally titled; it had deviated into a discussion about a crossplane integration but i think it's more general.

above @vincepri wrote:
> IIRC (although this was long ago) we've only used controller-type owner references where the lifecycle of that resource is managed by a Cluster API controller, while on the other resources we add the Cluster owner reference (or the parent) to make sure we have a way to tie things up together.

from looking at the code i think this is almost correct (granted it was a few years ago!) except for one thing: when adding the Cluster owner reference to it *is* actually a controller type ref that's being added. see here: https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/cluster/cluster_controller_phases.go#L95-L98 - this is when reconciling objects such as `AWSCluster`, it adds a controller ref linking it to the `Cluster` resource. later, in the AWS provider, it checks for that link in order to start doing the infra work: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controllers/awscluster_controller.go#L146 - if you dig into that function call you will see that it only checks for an ownerReference, not a controller ref.

i would like to see cluster-api only create controller refs for resources it creates (e.g. the AWS provider creating machines from a machine deployment), but use only owner refs for everything else where it needs to tie "things up together".

my use case, for context, is very similar to the OP except instead of crossplane it's metacontroller. it's [best-practice](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/controller-ref.md#the-three-laws-of-controllers) to take a controller ref of resources your controller creates and that's what metacontroller and crossplane are doing. in my case i am creating the following resources from a metacontroller operator: cluster, aws cluster, machine template, machine deployment, k3s controlplane, k3s config. metacontroller takes a controller ref which means that capi can't take one and the AWS provider won't start the infra work. i forked metacontroller to make setting controller: true in the ownerReference configurable and it works, but i think the problem is actually cluster-api side and i don't expect them to take that PR.

**Comment by sbueringer** [comment]
Sounds reasonable to me at a first glance.

I think we would have to differentiate between "standalone" clusters and clusters using clusterclass. With clusterclass we are actually creating infra cluster, control plane and MachineDeployments as well

**Comment by sbueringer** [comment]
cc @killianmuldoon given recent work on ownerRefs

**Comment by enxebre** [comment]
Revisiting this seems reasonable to me. IIRC The rationale behind using controller: true originally was mainly to prevent multiple Clusters from fighting over the "descendent" resources e.g MachineDep.

Also pending having better docs https://github.com/kubernetes-sigs/cluster-api/issues/5487

**Comment by fabriziopandini** [comment]
Just another data point on this; the code base is full of assumptions about OwnerReferences to be in place, and unfortunately, there is not good documentation linking owners with those assumptions (AFAIK the only exception is https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain that explicitly documents why it is relying on this info).

As a consequence changing owner reference should not be taken lightly, because it can lead to unexpected problems all around. We should figure out how to prevent regressions and how to communicate the change to all the Cluster API consumers that could have similar assumptions on their code
